<a href="https://colab.research.google.com/github/MathanRajOlaganathan/Coursera_Capstone/blob/master/Segmenting_and_Clustering_Neighborhoods_in_toronto.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
# importing libraries
import pandas as pd
import numpy as np
import re
import json
from pandas.io.json import json_normalize

Read the data from the following wiki page
https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [0]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [0]:
table_list = pd.read_html(url)

In [10]:
len(table_list)

3

The wiki page has 3 tables and the first table has list of postal codes in canada.Lets check the first table.

In [14]:
table_list[0][:5]

Postal code           Borough                Neighborhood
0         M1A      Not assigned                         NaN
1         M2A      Not assigned                         NaN
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

Lets convert the list in to dataframe

In [0]:
df = pd.DataFrame(table_list[0])

In [16]:
df.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                         NaN
1         M2A      Not assigned                         NaN
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

In [18]:
df.shape

(180, 3)

The inital dataframe has 180 observations

Lets remove the rows for which the  borough  is Not assigned.

In [0]:
df_rem = df[df['Borough']!='Not assigned'].reset_index(drop=True)

In [36]:
df_rem.head()

Postal code           Borough                                  Neighborhood
0         M3A        North York                                     Parkwoods
1         M4A        North York                              Victoria Village
2         M5A  Downtown Toronto                    Regent Park / Harbourfront
3         M6A        North York             Lawrence Manor / Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [37]:
df_rem.shape

(103, 3)

After removing the not assigned borough, there are 103 observations

Lets replace / with comma in Neighborhood column

In [0]:
df_rem['Neighborhood'] = df_rem['Neighborhood'].map(lambda x: re.sub(r"\s[/]", ",", x))


In [44]:
df_rem.shape

(103, 3)

Lets rename the Postal code column name

In [0]:
df_rem.rename(columns={'Postal code':'PostalCode'},inplace=True)

In [49]:
df_rem.columns


Index(['PostalCode', 'Borough', 'Neighborhood'], dtype='object')

In [50]:
df_rem.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [54]:
df_rem[df_rem['PostalCode']=='M5G']

PostalCode           Borough        Neighborhood
24        M5G  Downtown Toronto  Central Bay Street

In [51]:
df_rem['PostalCode'].value_counts()

M1R    1
M4V    1
M7R    1
M4R    1
M4K    1
      ..
M6B    1
M7Y    1
M9P    1
M9N    1
M1K    1
Name: PostalCode, Length: 103, dtype: int64

In [52]:
df_rem.shape

(103, 3)

In [0]:
43.69,-79.311

In [101]:
!pip install pgeocode

**Problem 2 -  get the latitude and the longitude coordinates of each neighborhood**

In [0]:
 import pgeocode

nomi = pgeocode.Nominatim('ca')

In [0]:
Latitude=[]
Longitude=[]
for i in range(df_rem.shape[0]):
  # address=str(df_rem['PostalCode'][i] +" "+ df_rem['Borough'][i] + " "+ 'canada')
  try:
    Latitude.append(nomi.query_postal_code(df_rem['PostalCode'][i]).latitude)
    Longitude.append(nomi.query_postal_code(df_rem['PostalCode'][i]).longitude)
  except (AttributeError,RuntimeError, TypeError, NameError):
    print('error')

In [113]:
len(Latitude),len(Longitude)

(103, 103)

In [0]:
df_rem['Latitude']=Latitude

In [0]:
df_rem['Longitude']=Longitude

In [119]:
df_rem.head()

PostalCode           Borough  ... Latitude  Longitude
0        M3A        North York  ...  43.7545   -79.3300
1        M4A        North York  ...  43.7276   -79.3148
2        M5A  Downtown Toronto  ...  43.6555   -79.3626
3        M6A        North York  ...  43.7223   -79.4504
4        M7A  Downtown Toronto  ...  43.6641   -79.3889

[5 rows x 5 columns]

**Problem3 - Explore and cluster the neighborhoods in Toronto.**

In [0]:
import requests
import folium
from sklearn.cluster import KMeans

Filter the Dataframe to include only the boroughs contain toronto

In [0]:
toronto_df = df_rem[df_rem['Borough'].str.contains('Toronto')].reset_index(drop=True)

In [133]:
toronto_df.head()

PostalCode           Borough  ... Latitude  Longitude
0        M5A  Downtown Toronto  ...  43.6555   -79.3626
1        M7A  Downtown Toronto  ...  43.6641   -79.3889
2        M5B  Downtown Toronto  ...  43.6572   -79.3783
3        M5C  Downtown Toronto  ...  43.6513   -79.3756
4        M4E      East Toronto  ...  43.6784   -79.2941

[5 rows x 5 columns]

FourSqaure Credentials

In [120]:
CLIENT_ID = 'FEVOBQA4RKSCY1YIPFWXESFZ530X3TL03CL0QKL4SACUY450' # your Foursquare ID
CLIENT_SECRET = '5FAMDZNCGGBNFPTGDLXMWAX3PE30FZ5GGYIARJHJF153D3AL' # your Foursquare Secret
VERSION = '20200406'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FEVOBQA4RKSCY1YIPFWXESFZ530X3TL03CL0QKL4SACUY450
CLIENT_SECRET:5FAMDZNCGGBNFPTGDLXMWAX3PE30FZ5GGYIARJHJF153D3AL


Function to explore the neighborhood

In [0]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [134]:
toronto_venues = getNearbyVenues(names=toronto_df['Neighborhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst  Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town

In [135]:
toronto_venues.head()

Neighborhood  ...  Venue Category
0  Regent Park, Harbourfront  ...     Coffee Shop
1  Regent Park, Harbourfront  ...          Bakery
2  Regent Park, Harbourfront  ...  Breakfast Spot
3  Regent Park, Harbourfront  ...             Spa
4  Regent Park, Harbourfront  ...  Breakfast Spot

[5 rows x 7 columns]

Let's check how many venues were returned for each neighborhood

In [136]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  ...  Venue Category
Neighborhood                                                               ...                
Berczy Park                                                            30  ...              30
Brockton, Parkdale Village, Exhibition Place                           30  ...              30
Business reply mail Processing CentrE                                  15  ...              15
CN Tower, King and Spadina, Railway Lands, Harb...                     30  ...              30
Central Bay Street                                                     30  ...              30
Christie                                                               12  ...              12
Church and Wellesley                                                   30  ...              30
Commerce Court, Victoria Hotel                                         30  ...              30
Davisville                                                             23  ...              23
Davisville North                                                        7  ...               7
Dufferin, Dovercourt Village                                           17  ...              17
First Canadian Place, Underground city                                 30  ...              30
Forest Hill North & West                                                3  ...               3
Garden District, Ryerson                                               30  ...              30
Harbourfront East, Union Station, Toronto Islands                       4  ...               4
High Park, The Junction South                                           2  ...               2
India Bazaar, The Beaches West                                         20  ...              20
Kensington Market, Chinatown, Grange Park                              30  ...              30
Lawrence Park                                                           2  ...               2
Little Portugal, Trinity                                               30  ...              30
Moore Park, Summerhill East                                             5  ...               5
North Toronto West                                                      4  ...               4
Parkdale, Roncesvalles                                                 30  ...              30
Queen's Park, Ontario Provincial Government                            27  ...              27
Regent Park, Harbourfront                                              23  ...              23
Richmond, Adelaide, King                                               30  ...              30
Rosedale                                                                4  ...               4
Roselawn                                                                1  ...               1
Runnymede, Swansea                                                     30  ...              30
St. James Town                                                         30  ...              30
St. James Town, Cabbagetown                                            30  ...              30
Stn A PO Boxes                                                         30  ...              30
Studio District                                                         8  ...               8
Summerhill West, Rathnelly, South Hill, Forest ...                      6  ...               6
The Annex, North Midtown, Yorkville                                    25  ...              25
The Beaches                                                             7  ...               7
The Danforth West, Riverdale                                           30  ...              30
Toronto Dominion Centre, Design Exchange                               30  ...              30
University of Toronto, Harbord                                         30  ...              30

[39 rows x 6 columns]

In [138]:
print('There are {} unique venue categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 173 unique venue categories.


Lets apply one hot encoding to venu categories

In [171]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
features = list(toronto_onehot.columns)
features.remove('Neighborhood')
fixed_columns = ['Neighborhood'] + features
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  American Restaurant  ...  Wings Joint  Yoga Studio
0  Regent Park, Harbourfront                    0  ...            0            0
1  Regent Park, Harbourfront                    0  ...            0            0
2  Regent Park, Harbourfront                    0  ...            0            0
3  Regent Park, Harbourfront                    0  ...            0            0
4  Regent Park, Harbourfront                    0  ...            0            0

[5 rows x 173 columns]

In [173]:
toronto_onehot.shape

(785, 173)

let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [174]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  ...  Yoga Studio
0                                         Berczy Park  ...     0.000000
1        Brockton, Parkdale Village, Exhibition Place  ...     0.000000
2               Business reply mail Processing CentrE  ...     0.066667
3   CN Tower, King and Spadina, Railway Lands, Har...  ...     0.033333
4                                  Central Bay Street  ...     0.000000
5                                            Christie  ...     0.000000
6                                Church and Wellesley  ...     0.000000
7                      Commerce Court, Victoria Hotel  ...     0.000000
8                                          Davisville  ...     0.000000
9                                    Davisville North  ...     0.000000
10                       Dufferin, Dovercourt Village  ...     0.000000
11             First Canadian Place, Underground city  ...     0.000000
12                           Forest Hill North & West  ...     0.000000
13                           Garden District, Ryerson  ...     0.000000
14  Harbourfront East, Union Station, Toronto Islands  ...     0.000000
15                      High Park, The Junction South  ...     0.000000
16                     India Bazaar, The Beaches West  ...     0.000000
17          Kensington Market, Chinatown, Grange Park  ...     0.000000
18                                      Lawrence Park  ...     0.000000
19                           Little Portugal, Trinity  ...     0.033333
20                        Moore Park, Summerhill East  ...     0.000000
21                                 North Toronto West  ...     0.000000
22                             Parkdale, Roncesvalles  ...     0.000000
23        Queen's Park, Ontario Provincial Government  ...     0.000000
24                          Regent Park, Harbourfront  ...     0.043478
25                           Richmond, Adelaide, King  ...     0.000000
26                                           Rosedale  ...     0.000000
27                                           Roselawn  ...     0.000000
28                                 Runnymede, Swansea  ...     0.000000
29                                     St. James Town  ...     0.000000
30                        St. James Town, Cabbagetown  ...     0.000000
31                                     Stn A PO Boxes  ...     0.000000
32                                    Studio District  ...     0.000000
33  Summerhill West, Rathnelly, South Hill, Forest...  ...     0.000000
34                The Annex, North Midtown, Yorkville  ...     0.000000
35                                        The Beaches  ...     0.000000
36                       The Danforth West, Riverdale  ...     0.033333
37           Toronto Dominion Centre, Design Exchange  ...     0.000000
38                     University of Toronto, Harbord  ...     0.033333

[39 rows x 173 columns]

In [175]:
toronto_grouped.shape

(39, 173)

Function to sort the venues in descending order.

In [0]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

let's create the new dataframe and display the top 10 venues for each neighborhood.

In [177]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  ... 10th Most Common Venue
0                                        Berczy Park  ...                 Lounge
1       Brockton, Parkdale Village, Exhibition Place  ...              Pet Store
2              Business reply mail Processing CentrE  ...    Japanese Restaurant
3  CN Tower, King and Spadina, Railway Lands, Har...  ...   Caribbean Restaurant
4                                 Central Bay Street  ...     Chinese Restaurant

[5 rows x 11 columns]

Cluster Neighborhoods

In [178]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 0, 2, 2, 2, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [180]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode  ... 10th Most Common Venue
0        M5A  ...     Italian Restaurant
1        M7A  ...    Distribution Center
2        M5B  ...          Burrito Place
3        M5C  ...                 Hostel
4        M4E  ...    Dumpling Restaurant

[5 rows x 16 columns]

Lets create map for the clusters

In [0]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [189]:
address = 'Toronto, Ontario, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 43.6534817, -79.3839347.


In [190]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [191]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...       10th Most Common Venue
7   Downtown Toronto  ...                      Dog Run
9       West Toronto  ...         Gym / Fitness Center
10  Downtown Toronto  ...                      Dog Run
20   Central Toronto  ...  Eastern European Restaurant
21   Central Toronto  ...                      Dog Run
23   Central Toronto  ...                      Dog Run
29   Central Toronto  ...                   Donut Shop
33  Downtown Toronto  ...                      Dog Run

[8 rows x 12 columns]

**The above clusters are highlighted in red color consist of Park, Dog Run, 

In [192]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...  10th Most Common Venue
19  Central Toronto  ...                 Dog Run

[1 rows x 12 columns]

**Cluster 1 has only one neighborhood and has mix of fitness and restaurants.**

In [193]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...  10th Most Common Venue
0   Downtown Toronto  ...      Italian Restaurant
1   Downtown Toronto  ...     Distribution Center
2   Downtown Toronto  ...           Burrito Place
3   Downtown Toronto  ...                  Hostel
4       East Toronto  ...     Dumpling Restaurant
5   Downtown Toronto  ...                  Lounge
6   Downtown Toronto  ...      Chinese Restaurant
8   Downtown Toronto  ...             Pizza Place
11      West Toronto  ...     Japanese Restaurant
12      East Toronto  ...                     Pub
13  Downtown Toronto  ...             Pizza Place
14      West Toronto  ...               Pet Store
15      East Toronto  ...                     Pub
16  Downtown Toronto  ...              Steakhouse
17      East Toronto  ...       Electronics Store
24   Central Toronto  ...            Burger Joint
25      West Toronto  ...          Ice Cream Shop
26   Central Toronto  ...              Restaurant
27  Downtown Toronto  ...              Beer Store
28      West Toronto  ...                     Pub
30  Downtown Toronto  ...            Cocktail Bar
31   Central Toronto  ...              Donut Shop
32  Downtown Toronto  ...    Caribbean Restaurant
34  Downtown Toronto  ...                    Café
35  Downtown Toronto  ...             Pizza Place
36  Downtown Toronto  ...              Steakhouse
37  Downtown Toronto  ...          Ice Cream Shop
38      East Toronto  ...     Japanese Restaurant

[28 rows x 12 columns]

**Cluster 2 has many neighborhoods and it has mostly restaurants**

In [194]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...  10th Most Common Venue
18  Central Toronto  ...     Distribution Center

[1 rows x 12 columns]

**Cluster 3 is similar to cluster 1 has only one neighborhood and consist of everything from park to Disribution center**

In [201]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
22  West Toronto               4  ...               Dog Run    Distribution Center

[1 rows x 12 columns]

**Cluster 4 is similar to cluster 1 and 3 has only one neighborhood and consist of everything from park to Disribution center**

**Observations**

*   From the map, it is clear that the neighborhood which has top 5 venues as coffee shop, bar, restaurant, cafe are clustered together in sky blue color.
*   Cluster 2 is the perfect place to start a restuarant, cluster 0,3 and 4 are the ideal spots for park and cluster 1 for spa.
